dimod/include/dimod/iterators.h

```diff
@@ -0,0 +1,319 @@
+            : neighborhood_ptr_(nullptr), i_(0), pair_ptr_(nullptr) {}
+
+    ConstNeighborhoodIterator(const neighborhood_type* neighborhood,
+                              index_type i)
+            : neighborhood_ptr_(neighborhood), i_(i), pair_ptr_(nullptr) {
+        update_pair();
+    }
+
+    const reference operator*() const { return *pair_ptr_; }
+
+    const pointer operator->() const { return pair_ptr_; }
+
+    ConstNeighborhoodIterator<Bias, Index>& operator++() {
+        ++i_;
+        update_pair();
+        return *this;
+    }
+    ConstNeighborhoodIterator<Bias, Index>& operator--() {
+        --i_;
+        update_pair();
+        return *this;
+    }
+    ConstNeighborhoodIterator<Bias, Index> operator++(int) {
+        ConstNeighborhoodIterator<Bias, Index> r(*this);
+        ++i_;
+        update_pair();
+        return r;
+    }
+    ConstNeighborhoodIterator<Bias, Index> operator--(int) {
+        ConstNeighborhoodIterator<Bias, Index> r(*this);
+        --i_;
+        update_pair();
+        return r;
+    }
+
+    ConstNeighborhoodIterator<Bias, Index>& operator+=(int n) {
+        i_ += n;
+        update_pair();
+        return *this;
+    }
+    ConstNeighborhoodIterator<Bias, Index>& operator-=(int n) {
+        i_ -= n;
+        update_pair();
+        return *this;
+    }
+
+    ConstNeighborhoodIterator<Bias, Index> operator+(int n) const {
+        ConstNeighborhoodIterator<Bias, Index> r(*this);
+        return r += n;
+    }
+    ConstNeighborhoodIterator<Bias, Index> operator-(int n) const {
+        ConstNeighborhoodIterator<Bias, Index> r(*this);
+        return r -= n;
+    }
+
+    difference_type operator-(
+            ConstNeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ - r.i_;
+    }
+
+    bool operator<(ConstNeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ < r.i_;
+    }
+    bool operator<=(ConstNeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ <= r.i_;
+    }
+    bool operator>(ConstNeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ > r.i_;
+    }
+    bool operator>=(ConstNeighborhoodIterator<Bias, Index> const& r) const {
+        return i_ >= r.i_;
+    }
+    bool operator!=(const ConstNeighborhoodIterator<Bias, Index>& r) const {
+        return i_ != r.i_;
+    }
+    bool operator==(const ConstNeighborhoodIterator<Bias, Index>& r) const {
+        return i_ == r.i_;
+    }
+
+ private:
+    const neighborhood_type* neighborhood_ptr_;
+    index_type i_;
+
+    value_type* pair_ptr_;
+
+    void update_pair() {
+        if (pair_ptr_) {
+            delete pair_ptr_;
+        }
+        if (i_ >= 0 && i_ < (index_type)neighborhood_ptr_->size()) {
+            pair_ptr_ = new value_type(neighborhood_ptr_->neighbors[i_],
+                                       neighborhood_ptr_->quadratic_biases[i_]);
+        }
+    }
+};
+
+template <class Bias, class Index>
+class ConstQuadraticIterator {
```

Comment:
The workings of ConstQuadraticIterator is much more complex than the NeighborhoodIterators.  It is better to add more elaborate comments to the advance function, with complete sentences.